src/helpers/buffers/fsv.rs

```diff
@@ -38,27 +41,39 @@ impl<const N: usize> FixedSizeByteVec<N> {
         Self {
             data: vec![0_u8; N * capacity],
             added: bitvec![0; capacity],
-            taken: 0,
+            capacity,
+            end: capacity,
         }
     }
 
     /// Inserts a new element to the specified position, returning the previous element at this `index`.
     /// ## Panics
     /// Panics if `index` is out of bounds or if something was previously inserted at `index`.
-    pub fn insert(&mut self, index: usize, elem: &[u8; N]) {
-        // if index is out of bounds, this line will panic, there is no need for additional check
-        // TODO: save runtime cost with `debug_assert!()`, though that would affects panic tests
-        assert!(!self.added[index]);
+    pub fn insert<D: Debug>(&mut self, channel: D, index: usize, elem: &[u8; N]) {
+        // Translate from an absolute index into a relative one.
+        let index = index
+            .checked_sub(self.end - self.capacity)
+            .unwrap_or_else(|| panic!("Duplicate send for index {index} on channel {channel:?}"));
         let start = index * N;
         let offset = start..start + N;
 
-        self.added.set(index, true);
+        assert!(
+            !self.added.replace(index, true),
+            "Duplicate send for index {index} on channel {channel:?}"
+        );
         self.data[offset].copy_from_slice(elem);
```

Comment:
It might be worth adding a comment here either explaining what it means when an OOB access occurs here, or referring to something that does (i.e. to make clear to the reader that the panic indicates erroneous usage of the buffer, and not a bug in the buffer implementation itself).